hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +78,945 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
     // sync the embedded group
     m_cl->setEmbeddedGroup(m_embed_group);
 
+    // check for collision warnings the first time collide is run
+    if (!m_checked_collision_warnings)
+        {
+        checkCollisionWarnings(timestep);
+        m_checked_collision_warnings = true;
+        }
+    if (m_needs_temperature && !m_T)
+        {
+        throw std::runtime_error("Temperature required by collision method");
+        }
+
+    // check the GPU autotuners
+#ifdef ENABLE_HIP
+    if (m_check_rigid_tuners)
+        {
+        checkRigidAutotuners();
+        m_check_rigid_tuners = false;
+        }
+#endif // ENABLE_HIP
+
     // set random grid shift
     m_cl->drawGridShift(timestep);
 
     // update cell list
     m_cl->compute(timestep);
 
+    // create auxillary array for rigid bodies
+    const bool rigid_body_collision = m_embed_group && m_rigid_bodies;
+    if (rigid_body_collision)
+        {
+        // resize initial velocity array
+        const unsigned int num_group = m_embed_group->getNumMembers();
+        if (num_group > m_initial_velocity.getNumElements())
+            {
+            GPUArray<Scalar4> initial_velocity(num_group, m_exec_conf);
+            m_initial_velocity.swap(initial_velocity);
+            }
+
+        const unsigned int num_total = m_pdata->getN() + m_pdata->getNGhosts();
+        if (num_total > m_linmom_accum.getNumElements())
+            {
+            GPUArray<Scalar3> linmom_accum(num_total, m_exec_conf);
+            m_linmom_accum.swap(linmom_accum);
+            GPUArray<Scalar3> angmom_accum(num_total, m_exec_conf);
+            m_angmom_accum.swap(angmom_accum);
+            }
+
+#ifdef ENABLE_HIP
+        if (m_exec_conf->isCUDAEnabled())
+            {
+            // thermalize constituent particles to add energy to the rigid body
+            thermalizeConstituentParticlesGPU(timestep);
+            // store constituent velocities
+            storeInitialEmbeddedGroupVelocitiesGPU(timestep);
+            }
+        else
+#endif
+            {
+            // thermalize constituent particles to add energy to the rigid body
+            thermalizeConstituentParticles(timestep);
+            // store constituent velocities
+            storeInitialEmbeddedGroupVelocities(timestep);
+            }
```

Comment:
A consolidated comment would be more helpful, especially since you have descriptive function names
```suggestion
        // thermalize rigid bodies and store constituent velocities
        if (m_exec_conf->isCUDAEnabled())
            {
            thermalizeConstituentParticlesGPU(timestep);
            storeInitialEmbeddedGroupVelocitiesGPU(timestep);
            }
        else
#endif
            {
            thermalizeConstituentParticles(timestep);
            storeInitialEmbeddedGroupVelocities(timestep);
            }
```